stan/math/fwd/fun/sum.hpp

```diff
@@ -34,25 +34,18 @@ inline fvar<T> sum(const std::vector<fvar<T> >& m) {
 /**
  * Return the sum of the entries of the specified matrix.
  *
- * @tparam T inner type of the fvar matrix
- * @tparam R number of rows, can be Eigen::Dynamic
- * @tparam C number of columns, can be Eigen::Dynamic
+ * @tparam T type of the matrix
  *
  * @param m Matrix.
  * @return Sum of matrix entries.
  */
-template <typename T, int R, int C>
-inline fvar<T> sum(const Eigen::Matrix<fvar<T>, R, C>& m) {
+template <typename T, require_eigen_vt<is_fvar, T>* = nullptr>
+inline value_type_t<T> sum(const T& m) {
```

Comment:
Having a trait metaprogram as return type tells someone who reads the code more than just `auto`. Since `value_type_t<T>` is not long and unreadable I prefere it to `auto`. 
But I agree that duplicating it in return statement does not make sense so I am changing the return statement to use implicit conversion.